Adds Resource azurerm_app_service_slot_virtual_network_swift_connection

cimnine opened the pull request:
Adds a new resource `azurerm_app_service_virtual_network_swift_connection_slot` which uses the [corresponding Azure API](https://docs.microsoft.com/en-us/rest/api/appservice/webapps/updateswiftvirtualnetworkconnectionslot) and is named after it.

Fixes #5458

katbyte commented on website/docs/r/app_service_virtual_network_swift_connection_slot.html.markdown:
````diff
@@ -0,0 +1,109 @@
+## Example Usage
+
+```hcl
+resource "azurerm_resource_group" "test" {
````
Could we use example here?
```suggestion
resource "azurerm_resource_group" "example" {
```

katbyte commented on website/docs/r/app_service_virtual_network_swift_connection_slot.html.markdown:
```diff
@@ -0,0 +1,109 @@
+  location = "uksouth"
+}
+
+resource "azurerm_virtual_network" "test" {
```
And for the rest
```suggestion
resource "azurerm_virtual_network" "example" {
```

katbyte commented on website/docs/r/app_service_virtual_network_swift_connection_slot.html.markdown:
```diff
@@ -0,0 +1,109 @@
+}
+
+resource "azurerm_virtual_network" "test" {
+  name                = "acctestvnet"
```
And not use test in resource names
```suggestion
  name                = "example-vnet"
```

cimnine commented:
Thank's for the early feedback.

> I've given this PR a quick review as its a draft (is it ready?) ...

Indeed, I'm still validating the proposed change, hence the _draft_ status of this PR.

> ... and left some comments inline that should be addressed. Additionally while running the tests i got: [error]

I have incorporated the feedback and fixed the test so far.

Except for the failing test – which was obviously a shortcoming on my side – would there have been any other way than the feedback to learn about these code conventions? I'm only asking because I did look out for a `CONTRIBUTION.md` or similar, but I have not succeeded in finding any relevant documentation.

cimnine commented:
The acceptance tests for the respective resource run on my machine, also the regular test suite.
The new resource `azurerm_app_service_virtual_network_swift_connection_slot` is modelled according to the existing resource `azurerm_app_service_virtual_network_swift_connection`.

An early idea was making the `azurerm_app_service_virtual_network_swift_connection` resource capable of doing both, slot and regular swift connections, but I think that the chosen solution – copying the existing provider and adjusting the relevant lines – yields code that is more maintainable and easier to comprehend.

cimnine commented:
Just because I was not sure that this is clear: From my point of view this PR is ready to be reviewed and eventually be merged.

cimnine commented:
Rebased onto latest master

katbyte commented on azurerm/internal/services/web/registration.go:
```diff
@@ -33,16 +33,17 @@ func (r Registration) SupportedDataSources() map[string]*schema.Resource {
+		"azurerm_app_service_slot":                                  resourceArmAppServiceSlot(),
+		"azurerm_app_service_source_control_token":                  resourceArmAppServiceSourceControlToken(),
+		"azurerm_app_service_virtual_network_swift_connection":      resourceArmAppServiceVirtualNetworkSwiftConnection(),
+		"azurerm_app_service_virtual_network_swift_connection_slot": resourceArmAppServiceVirtualNetworkSwiftConnectionSlot(),
```
could we make this 
```suggestion
		"azurerm_app_service_slot_virtual_network_swift_connection": resourceArmAppServiceVirtualNetworkSwiftConnectionSlot(),
```

cimnine replied:
I've named it this way is because when running the test just for the `azurerm_app_service_virtual_network_swift_connection` test, it would be impossible with the suggested naming, as both tests match.

cimnine replied:
Should I rename it anyway?

cimnine replied:
Oh. After reading this comment so many times I see that my argument was totally invalid and that you're even supporting it – kind of. A fix is incoming.

katbyte commented on azurerm/internal/services/web/resource_arm_app_service_virtual_network_swift_connection_slot.go:
```diff
@@ -0,0 +1,216 @@
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["sites"]
+	slotName := id.Path["slots"]
```
Could we write a parse ID function like can be seen in #5779

cimnine replied:
done

cimnine replied:
I've also implemented the change for `resource_arm_app_service_virtual_network_switch_connect.go`.

katbyte commented on azurerm/internal/services/web/resource_arm_app_service_virtual_network_swift_connection_slot.go:
```diff
@@ -0,0 +1,216 @@
+		}
+		return fmt.Errorf("Error retrieving existing App Service Slot %q (App Service %q / Resource Group %q): %s", slotName, name, resourceGroup, err)
+	}
+	appService, err := client.Get(ctx, resourceGroup, name)
```
```suggestion

	appService, err := client.Get(ctx, resourceGroup, name)
```

katbyte commented on azurerm/internal/services/web/resource_arm_app_service_virtual_network_swift_connection_slot.go:
```diff
@@ -0,0 +1,216 @@
+		}
+		return fmt.Errorf("Error retrieving existing App Service %q (Resource Group %q): %s", name, resourceGroup, err)
+	}
+	swiftVnet, err := client.GetSwiftVirtualNetworkConnectionSlot(ctx, resourceGroup, name, slotName)
```
```suggestion

	swiftVnet, err := client.GetSwiftVirtualNetworkConnectionSlot(ctx, resourceGroup, name, slotName)
```

katbyte commented on azurerm/internal/services/web/resource_arm_app_service_virtual_network_swift_connection_slot.go:
```diff
@@ -0,0 +1,216 @@
+	resourceGroup := id.ResourceGroup
+	name := id.Path["sites"]
+	subnetName := subnetID.Path["subnets"]
+	virtualNetworkName := subnetID.Path["virtualNetworks"]
```
We should be able to use d.ID to delete the resource?

cimnine replied:
I have copied this from the `resource_arm_app_service_virtual_network_switch_connect.go`. But of course. I also changed it in the other file.

katbyte commented on azurerm/internal/services/web/tests/resource_arm_app_service_slot_virtual_network_association_test.go:
```diff
@@ -0,0 +1,281 @@
+		ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
+
+		// Ensure we have enough information in state to look up in API
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Not found: %s", resourceName)
+		}
+
+		err, name, slotName, resourceGroup := parseResourceId(rs)
+		if err != nil {
+			return err
+		}
```
we can ue the new parse id function here

cimnine replied:
Yes.

katbyte commented on azurerm/internal/services/web/tests/resource_arm_app_service_slot_virtual_network_association_test.go:
```diff
@@ -0,0 +1,281 @@
+		client := acceptance.AzureProvider.Meta().(*clients.Client).Web.AppServicesClient
+		ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
+
+		// Ensure we have enough information in state to look up in API
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Not found: %s", resourceName)
+		}
+
+		err, name, slotName, resourceGroup := parseResourceId(rs)
+		if err != nil {
+			return err
+		}
```
we can ue the new parse id function here

cimnine replied:
yes

cimnine commented:
I have implemented all but one requested change, for which I've made my case. The branch is also rebased onto the latest master.

cimnine commented:
Rebased onto master.

cimnine commented:
Rebased onto `master` and renames resource to `azurerm_app_service_slot_virtual_network_swift_connection` as suggested by @katbyte.

JSkimming commented:
Just wondering where we are with this. Judging by the replies it's awaiting your review @katbyte

schlbra commented:
Any updates on this @katbyte ?

cimnine commented:
IMO it's ready to merge. I've just rebased onto `master` again to resolve the merge conflict.

cimnine commented on azurerm/internal/services/web/parse/slot_virtual_network_swift_connection.go:
```diff
@@ -7,23 +7,23 @@ import (
 )
 
 type SlotVirtualNetworkSwitchConnectionId struct {
```
Is it intentional that you did not rename the actual type's name as well? It seems to contain the same typo.

cimnine replied:
I'll just do it :)

jackofallops replied:
Nope, missed it, good catch! 😄

jackofallops commented:
Tests passing
![image](https://user-images.githubusercontent.com/11830746/86568250-7d6aaf80-bf64-11ea-92a9-977c97d4995b.png)

cimnine commented:
Thanks for merging 🥳
